### DISCUSSION TITLE
Better ergonomics for inline @Shared updates

### DISCUSSION THREAD

**Discussion by jegnux** [pointfreeco/swift-composable-architecture#3102] [Ideas]
If we want to modify the value of a shared state "inline" (from a function) for test or previews, we have to go through this syntax:
```swift
func testFeature() {
  @Shared(.appStorage("count")) var count = 42
  count = 42  // NB: Set again to override any value set by the app target.


  // Shared state will be 42 for all features using it.
  let store = TestStore(…)
}
```

And if we want to restore the default value of a `PersistenceKeyDefault` there's no real way to do it as https://github.com/pointfreeco/swift-composable-architecture/blob/8a40e932cf4f5827ced0fdfc1143aebf578dcbe3/Sources/ComposableArchitecture/SharedState/PersistenceKey/PersistenceKeyDefault.swift#L27
is internal. 

I would like to introduce 3 new APIs to provide better ergonomics to Shared:
```swift
extension Shared {
    public static func reset<T: PersistenceKey>(_ key: PersistenceKeyDefault<T>) where T.Value == Value {
        Self.set(key.defaultValue(), for: key)
    }

    public static func set<T: PersistenceKey>(_ value: Value, for key: PersistenceKeyDefault<T>) where T.Value == Value {
        Self.update(key) {
            $0 = value
        }
    }
    
    public static func update<Key: PersistenceKey>(_ key: Key, handler: (inout Key.Value) -> Void) where Key.Value == Value {
        guard let shared = try? Shared(key) else {
            return
        }
        handler(&shared.wrappedValue)
    }
}
```

So we can write:
```swift
func testFeature() {
  Shared.update(.appStorage("count")) {
    $0 = 42
  }

  // Shared state will be 42 for all features using it.
  let store = TestStore(…)
}
```

Or:
```swift
func testFeature() {
  Shared.set(42, for: .appStorage("count"))

  // Shared state will be 42 for all features using it.
  let store = TestStore(…)
}
```

Or:
```swift
extension PersistenceKey where Self == PersistenceKeyDefault<AppStorageKey<Int>> {
    public static var sharedCount: Self {
        .init(.appStorage("some_shared_count"), 42)
    }
}
```
```swift
func testFeature() {
  Shared.reset(.sharedCount)

  // Shared state will be 42 for all features using it.
  let store = TestStore(…)
}
```

If the team/community doesn't think it improves the Shared api ergonomics enough to add this to the library, can we at least make `PersistenceKeyDefault.defaultValue` public, so I (or anyone) can write these extensions in my project?

**Comment by jegnux** [comment]
Of course, if we introduce this in the library, we can consider other naming or error management (update func currently `try?` but we can rethrows the error)

**Comment by jegnux** [comment]
An alternative could be to add a `reset()` requirement to PersistenceKey so all keys can have a way to reset themselves (delete file, remove user defaults, etc)

**Comment by stephencelis** [comment]
@jegnux Generally in tests and previews you should not have to reset shared state so long as you follow this documented gotcha: https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/sharingstate#Previews

And so your first example becomes simply:

```swift
func testFeature() {
  @Shared(.appStorage("count")) var count = 42

  // Shared state will be 42 for all features using it.
  let store = TestStore(…)
}
```

And as such there should also be no need to call `reset`.

If you prefer the ergonomics of your helpers, you can totally define them in your own projects, and if so we could consider making `defaultValue` public, but I think we'd like to keep the library surface area a bit leaner. What do you think?

**Reply by jegnux** [reply]
```swift
@Shared(.appStorage("count")) var count = 42
```
I personnaly don't like it much because it uses the "default" syntax to set the value and rely on an implementation detail that I don't want to care about in tests, previews or elsewhere.

Also, my use case is beyond tests and previews, as I build a **Debug Sheet** in my application where I want to be able to inspect and reset my Shared states (see below, although for now it only just reset everything)

Actually, it would be really great for my use case to add a `reset()` requirement on all `PersistenceKey` because right now I'm only able to reset a `PersistenceKeyDefault` to its default, but the underlying key is not reset, it only replaces its current value with the default value which is saved/persisted on disk/user defaults/etc. With a `reset()` requirement I would have a real `reset` behavior on any key. But T can accommodate having only a public `defaultValue` for now.

<img src="https://github.com/pointfreeco/swift-composable-architecture/assets/63070/31aa37c4-416b-4f24-8f13-5e6e7b7216de" width="375"/>

**Reply by stephencelis** [reply]
@jegnux Thanks for the context!

> rely on an implementation detail that I don't want to care about in tests, previews or elsewhere

I don't think we consider this an "implementation detail" since it's not some internal thing, but just the way the feature is expected to and must behave.

Your debug sheet is a cool example, but I think the lack of public `defaultValue` is the only thing preventing you from implementing it, correct? As mentioned earlier, we'd definitely be open to making this public, or at least `@_spi`-available in some capacity for now.

**Reply by jegnux** [reply]
> As mentioned earlier, we'd definitely be open to making this public

Cool! What's the process? Can I simply open a PR?

**Reply by stephencelis** [reply]
Sure, let's start with that!

**Comment by jegnux** [comment]
Oops, I missed last answer from Stephen, but I can close the discussion as `defaultValue` has been made public in https://github.com/pointfreeco/swift-sharing/pull/89
